m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -143,19 +155,40 @@ private M3Reporter(Builder builder) {
         bucketValueTagKey = builder.histogramBucketName;
         bucketValFmt = String.format("%%.%df", builder.histogramBucketTagPrecision);
 
-        metricQueue = new LinkedBlockingQueue<>(builder.maxQueueSize);
+        queue = new LinkedBlockingQueue<>(builder.maxQueueSize);
+
+        ThreadFactory namedThreadFactory = createThreadFactory();
 
-        executor = builder.executor != null ? builder.executor : Executors.newFixedThreadPool(NUM_PROCESSORS);
+        executorService = builder.executor != null ? builder.executor : Executors.newFixedThreadPool(NUM_PROCESSORS, namedThreadFactory);
+        scheduledExecutorService = Executors.newSingleThreadScheduledExecutor(namedThreadFactory);
 
         clock = Clock.systemUTC();
 
         commonTags = builder.metricTagSet;
 
+        protocolFactory = thriftProtocolFactory;
+
         processorsShutdownLatch = new CountDownLatch(NUM_PROCESSORS);
 
-        processors = new ArrayList<>();
+        collectorEndpointSockedAddresses = builder.endpointSocketAddresses;
+        processors = new Processor[NUM_PROCESSORS];
+
         for (int i = 0; i < NUM_PROCESSORS; ++i) {
-            processors.add(bootProcessor(builder.endpointSocketAddresses));
+            processors[i] = bootProcessor(collectorEndpointSockedAddresses);
+        }
+
+        // Schedule regular heartbeat up-keeping processors up and running
+        scheduledExecutorService.scheduleAtFixedRate(this::heartbeat, 1, 1, TimeUnit.SECONDS);
```

Comment:
is every second heartbeat a bit too aggressive?

Reply:
It does very simple checks